vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+		defer lock.RUnlock()
+		return i.oidcCache.config, nil
+	}
+	lock.RUnlock()
+
+	var c oidcConfig
+	entry, err := s.Get(ctx, oidcConfigStorageKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if entry != nil {
+		if err := entry.DecodeJSON(&c); err != nil {
+			return nil, err
+		}
+	}
+
+	c.effectiveIssuer = c.Issuer
+	if c.effectiveIssuer == "" {
+		c.effectiveIssuer = i.core.redirectAddr + issuerPath
+	}
+
+	lock.Lock()
+	i.oidcCache.config = &c
+	lock.Unlock()
+
+	return &c, nil
+}
+
+// handleOIDCCreateKey is used to create a new named key or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateKey(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var key namedKey
+	entry, err := req.Storage.Get(ctx, namedKeyConfigPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry != nil {
+		if err := entry.DecodeJSON(&key); err != nil {
+			return nil, err
+		}
+	} else {
+		key.Name = name
+	}
+
+	if rotationPeriodRaw, ok := d.GetOk("rotation_period"); ok {
+		key.RotationPeriod = time.Duration(rotationPeriodRaw.(int)) * time.Second
+	} else if req.Operation == logical.CreateOperation {
+		key.RotationPeriod = time.Duration(d.Get("rotation_period").(int)) * time.Second
+	}
+
+	if verificationTTLRaw, ok := d.GetOk("verification_ttl"); ok {
+		key.VerificationTTL = time.Duration(verificationTTLRaw.(int)) * time.Second
+	} else if req.Operation == logical.CreateOperation {
+		key.VerificationTTL = time.Duration(d.Get("verification_ttl").(int)) * time.Second
+	}
+
+	if key.RotationPeriod < 1*time.Minute {
+		return logical.ErrorResponse("rotation_period must be at least one minute"), nil
+	}
+
+	if key.VerificationTTL > 10*key.RotationPeriod {
+		return logical.ErrorResponse("verification_ttl cannot be longer than 10x rotation_period"), nil
+	}
+
+	if algorithm, ok := d.GetOk("algorithm"); ok {
+		key.Algorithm = algorithm.(string)
+	} else if req.Operation == logical.CreateOperation {
+		key.Algorithm = d.Get("algorithm").(string)
+	}
+
+	if key.Algorithm != "RS256" {
+		return logical.ErrorResponse("unknown signing algorithm %q", key.Algorithm), nil
+	}
+
+	// Update next rotation time if it is unset or now earlier than previously set.
+	nextRotation := time.Now().Add(key.RotationPeriod)
+	if key.NextRotation.IsZero() || nextRotation.Before(key.NextRotation) {
+		key.NextRotation = nextRotation
+	}
+
+	// generate keys if creating a new key
+	if key.SigningKey == nil {
+		signingKey, err := generateKeys(key.Algorithm)
+		if err != nil {
+			return nil, err
+		}
+
+		key.SigningKey = signingKey
+		key.KeyRing = append(key.KeyRing, &expireableKey{KeyID: signingKey.Public().KeyID})
+
+		if err := saveOIDCPublicKey(ctx, req.Storage, signingKey.Public()); err != nil {
+			return nil, err
+		}
+	}
+
+	// store named key
+	entry, err = logical.StorageEntryJSON(namedKeyConfigPath+name, key)
+	if err != nil {
+		return nil, err
```

Comment:
In this case, and the error case on Put, should probably delete the public key that was saved since it's now invalid.

Reply:
I changed the cache flush to a defer that will always run, even in an error case. With the empty cache, a reload of public keys is triggered. This will only keep referenced public keys (i.e. ones in a keyring), which this one won't be so it'll get deleted.